gdb/dwarf2/read.c

```diff
@@ -3662,33 +3707,155 @@ lookup_dwp_signatured_type (struct dwarf2_cu *cu, ULONGEST sig)
   return sig_entry;
 }
 
+struct fallback_data
+{
+  ULONGEST sig;
+  struct dwo_unit **dwo_entry;
+};
+
+/* Traversal function for reading type units from a fallback objfile with dwo
+   files.  Look for a signatured type in a dwo file and return the dwo_unit
+   entry in the INFO.  */
+
+static int
+process_dwo_file_for_fallback_signatured_type (void **slot, void *info)
+{
+  struct dwo_file *dwo_file = (struct dwo_file *) *slot;
+  struct fallback_data *data = (fallback_data *) info;
+  struct dwo_unit find_dwo_entry, *dwo_entry;
+
+  if (dwo_file->tus == NULL)
+    return 1;
+  find_dwo_entry.signature = data->sig;
+  dwo_entry = (struct dwo_unit *) htab_find (dwo_file->tus.get (),
+					     &find_dwo_entry);
+  if (dwo_entry == NULL)
+    return 1;
+
+  /* Stop traversal of dwo files */
+  *(data->dwo_entry) = dwo_entry;
+  return 0;
+}
+
+
+/* Subroutine of lookup_signatured_type.
+   Look up the type for signature SIG, from a fallback objfile.
+   If we can't find SIG in the fallback object file, check for dwo/dwp files. */
+
+static struct signatured_type *
+lookup_fallback_signatured_type (struct objfile *fallback_objfile, ULONGEST sig,
+				 struct dwarf2_per_objfile **per_objfile)
+{
+  struct signatured_type *sig_type = NULL;
+
+  if (fallback_objfile->separate_debug_objfile != NULL)
+    fallback_objfile = fallback_objfile->separate_debug_objfile;
+
+  struct dwarf2_per_objfile *fallback_per_objfile = get_dwarf2_per_objfile (
+	  fallback_objfile);
+
```

Comment:
Add comment:
First look into signature type of fallback per objfile if it has signatured types.